lib/Dialect/FIRRTL/LowerTypes.cpp

```diff
@@ -0,0 +1,485 @@
+LogicalResult FIRRTLTypesLowering::visitExpr(SubfieldOp op) {
+  Value input = op.input();
+  Value result = op.result();
+  StringRef fieldname = op.fieldname();
+
+  Value rootBundle;
+  SmallString<16> suffix;
+  if (input.isa<BlockArgument>()) {
+    // When the input is from a block argument, map the result to the root
+    // bundle block argument, and begin creating a suffix string.
+    rootBundle = input;
+    suffix.push_back('_');
+    suffix += fieldname;
+  } else if (input.isa<OpResult>()) {
+    Operation *owner = input.cast<OpResult>().getOwner();
+    if (isa<SubfieldOp>(owner)) {
+      // When the input is from a subfield, look up the lowered info from the
+      // parent.
+      auto subfieldLowering = getSubfieldLowering(input);
+      assert(subfieldLowering.hasValue() && "no subfield lowering for input");
+
+      // Map the result to the root bundle block argument, and append the
+      // suffix to what was already mapped.
+      auto subfieldInfo = subfieldLowering.getValue();
+      rootBundle = subfieldInfo.first;
+      suffix.assign(subfieldInfo.second);
+      suffix.push_back('_');
+      suffix.append(fieldname);
+    } else if (isa<InstanceOp>(owner)) {
+      // When the input is from an instance, look up the new instance.
+      Optional<Value> newInstance = getInstanceLowering(input);
+      assert(newInstance.hasValue() && "no instance lowering for input");
+
+      // Check if this subfield was originally a bundle type.
+      BundleType oldType = input.getType().cast<BundleType>();
+      FIRRTLType elementType = oldType.getElementType(fieldname);
+      assert(elementType && "no element found in lowered bundle");
+
+      if (elementType.isa<BundleType>()) {
+        // If it was a bundle, flatten it the usual way.
+        SmallVector<FlatBundleFieldEntry, 8> fieldTypes;
+        flattenBundleTypes(elementType, fieldname, false, fieldTypes);
+
+        for (auto field : fieldTypes) {
+          // Create a new subfield for the flattened field.
+          auto newSubfield =
+              builder->create<SubfieldOp>(op.getLoc(), getPortType(field),
+                                          newInstance.getValue(), field.suffix);
+
+          // Use the original result and the sub-field name (without the prefix)
+          // to remember the new subfield.
+          auto fieldName = StringRef(field.suffix).drop_front(fieldname.size());
+          setBundleLowering(result, fieldName, newSubfield);
+        }
+      }
+    }
+  }
+
+  Optional<Value> newValue = getBundleLowering(rootBundle, suffix);
+  if (newValue.hasValue()) {
+    // If the lowering for the current result exists, we have finished
+    // flattening. Replace the current result with the flattened value.
+    result.replaceAllUsesWith(newValue.getValue());
+  } else {
+    // Otherwise, remember the root bundle and the suffix so far.
+    setSubfieldLowering(result, rootBundle, std::string(suffix));
```

Comment:
Can you write a test for this 'else' code, or is it dead code?

Reply:
The recursive test exercises this: https://github.com/llvm/circt/pull/157/files#diff-cfcb11e78479b63cf6e909ee4a9e01a0ff1ba2ab019da32a0c64baa56b7e0c12R38. When we are lowering a subfield of a subfield, we come through this `else` branch, and continue building up the field suffix. In the base case, we hit the `if` branch with a complete suffix and look up the flattened bundle that was created for that suffix.